projects/angular-cesium/src/lib/angular-cesium/components/ac-map-terrain-provider/ac-map-terrian-provider.component.ts

```diff
@@ -0,0 +1,89 @@
+import { Component, OnInit, Input, OnChanges, SimpleChanges, OnDestroy } from '@angular/core';
+import { CesiumService } from '../../services/cesium/cesium.service';
+import { Checker } from '../../utils/checker';
+import { MapTerrainProviderOptions } from '../../models/map-terrain-provider-options.enum';
+
+/**
+ *  This component is used for adding a terrain provider service to the map (ac-map)
+ *  options according to selected terrain provider MapTerrainProviderOptions enum.
+ *
+ *
+ *  __Usage :__
+ *  ```
+ *    <ac-map-terrain-provider [options]="optionsObject" [provider]="myProvider">
+ *    </ac-map-terrain-provider>
+ *  ```
+ */
+@Component({
+  selector: 'ac-map-terrain-provider',
+  template: '',
+})
+export class AcMapTerrainProviderComponent implements OnInit, OnChanges, OnDestroy {
+
+  /**
+   * refer to cesium docs for details https://cesiumjs.org/Cesium/Build/Documentation/TerrainProvider.html
+   */
+  @Input()
+  options: { url?: string } = {};
+
+  /**
+   * the provider
+   */
+  @Input()
+  provider: any;
+
+  /**
+   * show (optional) - Determines if the map layer is shown.
+   */
+  @Input()
+  show = true;
+
+  private terrainProvider: any;
+  private defaultTerrainProvider: any;
+
+  constructor(private cesiumService: CesiumService) {
+  }
+
+  ngOnInit(): void {
+    if (!Checker.present(this.options.url) && this.provider !== MapTerrainProviderOptions.Ellipsoid && this.provider !== MapTerrainProviderOptions.WorldTerrain) {
+      throw new Error('options must have a url');
+    }
+    this.defaultTerrainProvider = this.cesiumService.getViewer().terrainProvider;
+    switch (this.provider) {
+      case MapTerrainProviderOptions.CesiumTerrain:
+      case MapTerrainProviderOptions.ArcGISTiledElevation:
+      case MapTerrainProviderOptions.GoogleEarthEnterprise:
+      case MapTerrainProviderOptions.VRTheWorld:
+      case MapTerrainProviderOptions.Ellipsoid:
+        this.terrainProvider = new this.provider(this.options);
+        break;
+      case MapTerrainProviderOptions.WorldTerrain:
+        this.terrainProvider = this.provider(this.options);
```

Comment:
Looks like world terrain and all other terrains have the same logic, so the switch case is redundant. Maybe you can change it to a simple if else

Reply:
The world terrain provider function already returns a new instance. The other terrain providers is like class that should have new.

Reply:
Indeed, I meant to use the `if` something like:
```
if (this.provider === MapTerrainProviderOptions.WorldTerrain) {
this.terrainProvider = this.provider(this.options);
} else {
// all others 
this.terrainProvider = new this.provider(this.options);
}
```

Reply:
But the switch case has default that has different case from the others, so If I will change it to if else I still need to check every provider option. In addition It maintains uniformity with the ac-map-layers-provider component that has the same behavior. If you are anyway prefer it in if else comment again and I will change it.